### PR TITLE
3.1 Support

### PR DESCRIPTION
**Pull request by GanyusLeftHorn**
## Description

This PR brings GC to 3.1! Well, not much changed this time around, but still. =D

Changes:
- New material types in the inventory.
- Add scene tags for the desert and the 3.1 main event area in Springvale.
- Update game version.

Thanks to all contributors in Grasscutter Backrooms for working on getting GC updated to new game versions so fast.

## Type of changes

- [x] Bug fix
- [x] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

### CODE REVIEW

File: src/main/java/emu/grasscutter/server/packet/send/PacketPlayerWorldSceneInfoListNotify.java [hunk 1]
```diff
@@ -63,18 +72,15 @@ public PacketPlayerWorldSceneInfoListNotify() {
                                 .setSceneId(7)
                                 .setIsLocked(false)
                                 .build()
+                )
+                .addInfoList(
+                        PlayerWorldSceneInfo.newBuilder()
+                                .setSceneId(9)
+                                .setIsLocked(false)
+                                .addAllSceneTagIdList(IntStream.range(0, 3000).boxed().collect(Collectors.toList()))
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketPlayerWorldSceneInfoListNotify.java]:
```suggestion
                                .addAllSceneTagIdList(IntStream.range(0, 3000).boxed().toList())
```

File: src/main/java/emu/grasscutter/server/packet/send/PacketPlayerWorldSceneInfoListNotify.java [hunk 2]
```diff
@@ -27,14 +24,26 @@ public PacketPlayerWorldSceneInfoListNotify() {
...
+
+                                // 3.1 event
+                                .addSceneTagIdList(152)
+                                .addSceneTagIdList(153)
+
+                                // Pyramid
+                                .addSceneTagIdList(1164) // Arena (XMSM_CWLTop)
+                                .addSceneTagIdList(1166) // Pyramid (CWL_Trans_02)
+
+                                // Brute forace
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketPlayerWorldSceneInfoListNotify.java]:
```suggestion
                                // Brute force
```

File: src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java [hunk 3]
```diff
@@ -61,9 +60,9 @@ public PacketPlayerEnterSceneNotify(Player player, Player target, EnterType type
                 .setWorldType(1)
                 .setSceneTransaction(newScene + "-" + target.getUid() + "-" + (int) (System.currentTimeMillis() / 1000) + "-" + 18402);
 
-        for (int i = 0; i < 3000; i++) {
+        /*for (int i = 0; i < 3000; i++) {
             proto.addSceneTagIdList(i);
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java]:
```suggestion
```

File: src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java [hunk 4]
```diff
@@ -61,9 +60,9 @@ public PacketPlayerEnterSceneNotify(Player player, Player target, EnterType type
...
-        }
+        }*/
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java]:
```suggestion
```

File: src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java [hunk 5]
```diff
@@ -19,7 +19,7 @@ public PacketPlayerEnterSceneNotify(Player player) {
         player.setSceneLoadState(SceneLoadState.LOADING);
         player.setEnterSceneToken(Utils.randomRange(1000, 99999));
 
-        PlayerEnterSceneNotify proto = PlayerEnterSceneNotify.newBuilder()
+        PlayerEnterSceneNotify.Builder proto = PlayerEnterSceneNotify.newBuilder()
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java:22]:
```suggestion
        PlayerEnterSceneNotify proto = PlayerEnterSceneNotify.newBuilder()
```

File: src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java [hunk 6]
```diff
@@ -30,8 +30,7 @@ public PacketPlayerEnterSceneNotify(Player player) {
                 .setEnterReason(EnterReason.Login.getValue())
                 .setIsFirstLoginEnterScene(player.isFirstLoginEnterScene())
                 .setWorldType(1)
-                .setSceneTransaction("3-" + player.getUid() + "-" + (int) (System.currentTimeMillis() / 1000) + "-" + 18402)
-                .build();
+                .setSceneTransaction("3-" + player.getUid() + "-" + (int) (System.currentTimeMillis() / 1000) + "-" + 18402);
```
Comment (Birdulon) [inline src/main/java/emu/grasscutter/server/packet/send/PacketPlayerEnterSceneNotify.java:33]:
```suggestion
                .setSceneTransaction("3-" + player.getUid() + "-" + (int) (System.currentTimeMillis() / 1000) + "-" + 18402)
                .build();
```

**Review by Birdulon — CHANGES_REQUESTED** [review]
Thanks everyone!

**Review by Birdulon — APPROVED** [review]
Squash it tomorrow when maintenance starts 👍

**Comment by Hartie95** [comment]
The new resources HomeWoldFurnitureExcelConfigData.json has the `roomSceneId` field deobfuscated, so  `emu.grasscutter.data.excels.ItemData` needs this small change to not trigger exceptions:
```java
@SerializedName(value="roomSceneId", alternate={"DANFGGLKLNO", "JFDLJGDFIGL", "OHIANNAEEAK"})
    private int roomSceneId;
```

and InvestigationMonsterConfigData.json has an entry with an empty groupIdList, so `emu.grasscutter.game.world.WorldDataSystem` will also need this additional check:
```java
private InvestigationMonsterOuterClass.InvestigationMonster getInvestigationMonster(Player player, InvestigationMonsterData imd) {
        if(imd.getGroupIdList().isEmpty() || imd.getMonsterIdList().isEmpty()){
            return null;
        }
```

**Comment by TukanDev** [comment]
wonderhoy

**Pull request merged**